### PR TITLE
[core] add Collapse component

### PR DESCRIPTION
**Pull request by EmilMalanczak**
Hi, as it is mentioned in the  [task ](https://trello.com/c/340F9YLe) I based on the collapse implementation from the Accordion component. I have tried to test the behaviour of height changes but had some problems with that. 

I believe I did not miss any checks from the component guidelines 😄 

Looking forward to a review.

### CODE REVIEW

File: docs/src/docs/core/Collapse.mdx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+styles: ['Collapse']
+---
+
+import { CollapseDemos } from '@demos/core';
+
+## Usage
+
+<Demo data={CollapseDemos.usage} />
+
+## Styles API
```
Comment (rtivital) [inline docs/src/docs/core/Collapse.mdx:22]: Collapse is a very simple component there is no need for Styles API demo and docs, remove it

File: src/mantine-core/src/components/Collapse/Collapse.story.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis
+        aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
+        pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia
+        deserunt mollit anim id est laborum.
+      </Collapse>
+    </div>
+  );
+});
+
+storiesOf('@mantine/core/Collapse', module).add('Initially opened', () => {
```
Comment (rtivital) [inline src/mantine-core/src/components/Collapse/Collapse.story.tsx:34]: Use single `storiesOf('@mantine/core/Collapse', module)` for all stories, see other components stories for an example

File: src/mantine-core/src/components/Collapse/Collapse.story.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    </div>
+  );
+});
+
+storiesOf('@mantine/core/Collapse', module).add('With transition API', () => {
+  const [open, setOpen] = useState(true);
+
+  const handleTransitionEnd = () => {
+    // eslint-disable-next-line no-alert
+    alert('transition end');
```
Comment (rtivital) [inline src/mantine-core/src/components/Collapse/Collapse.story.tsx:67]: Do not use alert, even in stories, it is disruptive, replace it with console.log if needed

File: src/mantine-core/src/components/Collapse/Collapse.styles.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { createMemoStyles, MantineTheme } from '../../theme';
+
+interface CollapseStyles {
+  theme: MantineTheme;
+  transitionDuration: number;
+  transitionTimingFunction: string;
+}
+
+export default createMemoStyles({
```
Comment (rtivital) [inline src/mantine-core/src/components/Collapse/Collapse.styles.ts:9]: There are too little styles for jss, make all styles inline (without jss)

File: src/mantine-core/src/components/Collapse/demos/accessibility.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        onClick={() => setOpen(prev => !prev)}
+        aria-expanded={isOpen}
+        aria-controls="collapse"
+        id="collapse-button"
+      >
+        Collapse
+      </Button>
+
+      <Collapse in={isOpen} aria-labelledby="collapse-button" id="collapse">
+        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
```
Comment (rtivital) [inline src/mantine-core/src/components/Collapse/demos/accessibility.tsx:20]: Do not include large sections of the content in demo code, replace it with comment `{/* content */}`

File: src/mantine-core/src/components/Collapse/demos/transition.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import React, { useState } from 'react';
+import { Button } from '../../Button/Button';
+import { Collapse } from '../Collapse';
+
+const code = `
+const [isOpen, setOpen] = useState(false);
```
Comment (rtivital) [inline src/mantine-core/src/components/Collapse/demos/transition.tsx:6]: If you use state, include whole Demo component in demo code

**Review by rtivital — CHANGES_REQUESTED** [review]
Thanks for PR, I have not yet reviewed component behavior, these comments are based on code only, I'll provide more feedback later if anything is wrong with component behavior

File: src/mantine-core/src/components/Collapse/Collapse.story.tsx (hunk 3 above)
Comment (EmilMalanczak) [reply src/mantine-core/src/components/Collapse/Collapse.story.tsx:67]: Shall I replace it in docs as well? There's an alert also

**Comment by EmilMalanczak** [comment]
Sure, I'll wait for a further review then I'll make the requested changed

File: src/mantine-core/src/components/Collapse/Collapse.story.tsx (hunk 3 above)
Comment (rtivital) [reply src/mantine-core/src/components/Collapse/Collapse.story.tsx:67]: yes

File: src/mantine-core/src/components/Collapse/Collapse.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    },
+    classNames,
+    'collapse'
+  );
+  const _styles = mergeStyles(classes, styles);
+  const [rect, setRect] = useState<HTMLDivElement>(null);
+
+  useWindowEvent('resize', forceUpdate);
+
+  const handleTransitionEnd = (e) => {
```
Comment (rtivital) [inline src/mantine-core/src/components/Collapse/Collapse.tsx:61]: Rename `e` -> `event`, add ts type

**Comment by rtivital** [comment]
I've checked how it works, everything looks fine.
The only thing that is not there yet is dynamic `children` handling, when children changes height should be recalculated (same as with resize event)

File: src/mantine-core/src/components/Collapse/Collapse.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+  useEffect(() => {
+    const content = contentRef?.current;
+
+    if (content) {
+      const { height: boundingHeight } = content.getBoundingClientRect();
+
+      setHeight(`${boundingHeight}px`);
+    }
+  }, [children]);
```
Comment (EmilMalanczak) [inline src/mantine-core/src/components/Collapse/Collapse.tsx:60]: The reason why I did not attach ref into dependency array - [link](https://epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array/)

**Comment by EmilMalanczak** [comment]
I've provided an additional storybook demo for dynamic children calculations

**Comment by rtivital** [comment]
Looks good, thanks for contrubution!

**Pull request merged**